bombastic/index/src/packages/mod.rs

```diff
@@ -0,0 +1,564 @@
+    }
+
+    fn index_cyclonedx(
+        &self,
+        bom: &cyclonedx_bom::prelude::Bom,
+        sha256: &str,
+    ) -> Result<Vec<(String, Document)>, SearchError> {
+        let mut documents: Vec<(String, Document)> = Vec::new();
+
+        if let Some(components) = &bom.components {
+            for component in components.0.iter() {
+                Self::index_cyclonedx_component(&mut documents, component, &self.fields, sha256);
+            }
+        }
+
+        Ok(documents)
+    }
+
+    fn index_cyclonedx_component(
+        documents: &mut Vec<(String, Document)>,
+        component: &cyclonedx_bom::prelude::Component,
+        fields: &Fields,
+        sha256: &str,
+    ) {
+        let mut document = doc!();
+        document.add_text(fields.sha256, sha256);
+        document.add_date(fields.indexed_timestamp, DateTime::from_utc(OffsetDateTime::now_utc()));
+        if let Some(hashes) = &component.hashes {
+            for hash in hashes.0.iter() {
+                if hash.alg == HashAlgorithm::SHA256 {
+                    document.add_text(fields.sha256, &hash.content.0);
+                }
+            }
+        }
+
+        document.add_text(fields.name, component.name.to_string());
+        document.add_text(fields.version, component.version.to_string());
+        let mut package_id = "".to_string();
+        if let Some(purl) = &component.purl {
+            let purl = purl.to_string();
+            package_id = purl.clone();
+            document.add_text(fields.purl, &purl);
+
+            if let Ok(package) = packageurl::PackageUrl::from_str(&purl) {
+                document.add_text(fields.purl_name, package.name());
+                if let Some(namespace) = package.namespace() {
+                    document.add_text(fields.purl_namespace, namespace);
+                }
+
+                if let Some(version) = package.version() {
+                    document.add_text(fields.purl_version, version);
+                }
+
+                for entry in package.qualifiers().iter() {
+                    document.add_text(fields.purl_qualifiers, entry.1);
+                }
+                document.add_text(fields.purl_type, package.ty());
+            }
+        }
+
+        if let Some(desc) = &component.description {
+            document.add_text(fields.desc, desc.to_string());
+        }
+
+        if let Some(licenses) = &component.licenses {
+            licenses.0.iter().for_each(|l| match l {
+                LicenseChoice::License(l) => match &l.license_identifier {
+                    LicenseIdentifier::Name(s) => {
+                        document.add_text(fields.license, s.to_string());
+                    }
+                    LicenseIdentifier::SpdxId(_) => (),
+                },
+                LicenseChoice::Expression(_) => (),
+            });
+        }
+
+        document.add_text(fields.classifier, component.component_type.to_string());
+        documents.push((package_id, document));
```

Comment:
This should only add the document if package_id is non-empty too I think.

Reply:
the id use in here https://github.com/trustification/trustification/blob/5f9b67b6bfe52b54b1f4f7afa2b1e0ee9110288d/index/src/lib.rs#L338. index will delete old doc before add new doc.

Reply:
Right, but if it didn't find a purl within the component above, the package_id would have the value `""`, which would be invalid.